### ISSUE TITLE
Conversion error

### ISSUE THREAD

**Issue by firepolo** [ConvertAPI/convertapi-php#56]
I noticed that the conversion didn't give the same result depending on the machine running the ConvertAPI PHP library. I found that this was due to the encoding when the file was uploaded.

I haven't investigated this in depth but it's a possibility. I managed to correct the problem by changing the header to ‘Transfer-Encoding: gzip,deflate’.

File:
`lib/ConvertApi/Client.php`
Line:
`30`
Original:
`‘Transfer-Encoding: chunked’,`
Fix:
`‘Transfer-Encoding: gzip,deflate’,`

**Comment by tomasr78** [comment]
@laurynas-convertapi, could you investigate and confirm, and if needed, apply the fix? Let me know If you need help from our internal developers.

**Comment by orlu-nd** [comment]
As discussed using the chat with Paulius and Tomas a while back and since we are (still) seeying the same issues as metioned by @firepolo, hereby more details like you asked @tomasr78.

Lately we have been experiencing a lot of error 5001 error codes while trying to convert documents (docx, xlsx, png, jpg) to pdf. We are using the following function:

```php
use ConvertApi\ConvertApi;
use ConvertApi\FileUpload;

$result = ConvertApi::convert('pdf', [
    'File' => new FileUpload($fileStream, $fileName),
    'StoreFile' => false,
    'Timeout' => 30,
]);
```

We thought we were delivering wrong/corrupt documents, since all API calls worked properly. However, we were unable to find the reason the error 5001 was triggered (unable to convert uploaded document). Even rewriting the above code, using files in stead of streams, got the same 5001 error. Contacting support, and uploading the documents manually resulted in the correct conversions. 

After doing some research, we found the issue in the [upload function](https://github.com/ConvertAPI/convertapi-php/blob/master/lib/ConvertApi/Client.php#L24), using the [FileUpload class](https://github.com/ConvertAPI/convertapi-php/blob/master/lib/ConvertApi/FileUpload.php).

On macOS we have written a simple test script:

```php
#!/opt/homebrew/opt/php@8.3/bin/php
<?php
require_once('./vendor/autoload.php');

use ConvertApi\ConvertApi;
use ConvertApi\Client;

ConvertApi::setApiSecret('your-api-secret');

$file = 'test.docx';

$fileContents = file_get_contents($file);
$crc = hash('crc32b', $fileContents);

echo "Original CRC: " . $crc . PHP_EOL;

$result = (new Client)->upload($file, $file);

echo "Upload url: " . $result['Url'] . PHP_EOL;

$crc = hash('crc32b', file_get_contents($result['Url']));
echo "Uploaded CRC: " . $crc . PHP_EOL;
```

To replicate, in an empty folder use `composer require convertapi/convertapi-php`, copy in the above test.php script (make it executable `chmod +x test.php`, also place the [test.docx](https://github.com/ConvertAPI/convertapi-php/blob/master/examples/files/test.docx) document (provided by the repository) in the same folder. 

We get the following result running the above script:

```bash
➜  temp ./test.php
Original CRC: 7937aa29
Upload url: https://v2.convertapi.com/d/2a0r9vequ6efsm5vsa1xdzf33nahgbfc
Uploaded CRC: 6319cecd
```

The [upload URL](https://v2.convertapi.com/d/2a0r9vequ6efsm5vsa1xdzf33nahgbfc) provides a corrupt file. Somewhere multiple bytes at the start of the file have been added, making it corrupt and therefore unable to convert.

Commenting out [line #30 of /ConvertApi/Client](https://github.com/ConvertAPI/convertapi-php/blob/master/lib/ConvertApi/Client.php#L30) solves the issue:

```bash
➜  temp ./test.php
Original CRC: 7937aa29
Upload url: https://v2.convertapi.com/d/sor8qc8zufhth05hjfpa459pexgpo51w
Uploaded CRC: 7937aa29
```

With [line #30 of /ConvertApi/Client](https://github.com/ConvertAPI/convertapi-php/blob/master/lib/ConvertApi/Client.php#L30) disabled, our original function worked properly again and the 5001 error disappeared.

Tested on:

PHP: 8.3.7 (attached a [phpinfo()](https://github.com/ConvertAPI/convertapi-php/files/15390350/phpinfo.log) export)
macOS: 14.4.1 (23E224)

We do not believe this is an issue of the PHP Client. However, somewhere the uploaded file is getting altered/corrupt using chunked Transfer-Encoding.

 Thanks in advance for looking in to this matter.

**Comment by tomasr78** [comment]
@orlu-nd Could you provide the original and corrupted files? Please post them here for debugging.

**Comment by orlu-nd** [comment]
➜  temp ./test.php
Original CRC: 7937aa29
Upload url: https://v2.convertapi.com/d/zp4b0kc380zeyxlj0aq1l2phso5bd0l9
Uploaded CRC: 6319cecd

Original file: [test.docx](https://github.com/ConvertAPI/convertapi-php/blob/master/examples/files/test.docx)
Corrupted/altered file (from the Upload url): [test(1).docx](https://github.com/user-attachments/files/15928503/test.1.docx)

**Comment by gwcms** [comment]
Hello, could you try replacing upload function in file lib/ConvertApi/Client.php
just replace function and please reply with test results

`    public function upload($file_or_resource, $fileName)
    {
        $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, []);


        if (!file_exists($file_or_resource))
        {
            throw new Error\File("Unable to open file {$file_or_resource}");
        }

        curl_setopt($ch, CURLOPT_POST, true);

        // Define the file to be uploaded
        $cfile = new \CURLFile($file_or_resource, mime_content_type($file_or_resource), basename($file_or_resource));

        // Attach the file to the POST fields
        $post_fields = array('file' => $cfile);
        curl_setopt($ch, CURLOPT_POSTFIELDS, $post_fields);

        // Set additional options
        curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
        curl_setopt($ch, CURLOPT_MAXREDIRS, 10);

        try
        {
            $result = $this->execute($ch);
        }
        catch (\Exception $e)
        {
            fclose($fp);
            throw $e;
        }

        return $result;
    }`


> As discussed using the chat with Paulius and Tomas a while back and since we are (still) seeying the same issues as metioned by @firepolo, hereby more details like you asked @tomasr78.
> 
> Lately we have been experiencing a lot of error 5001 error codes while trying to convert documents (docx, xlsx, png, jpg) to pdf. We are using the following function:
> 
> ```
> use ConvertApi\ConvertApi;
> use ConvertApi\FileUpload;
> 
> $result = ConvertApi::convert('pdf', [
>     'File' => new FileUpload($fileStream, $fileName),
>     'StoreFile' => false,
>     'Timeout' => 30,
> ]);
> ```
> 
> We thought we were delivering wrong/corrupt documents, since all API calls worked properly. However, we were unable to find the reason the error 5001 was triggered (unable to convert uploaded document). Even rewriting the above code, using files in stead of streams, got the same 5001 error. Contacting support, and uploading the documents manually resulted in the correct conversions.
> 
> After doing some research, we found the issue in the [upload function](https://github.com/ConvertAPI/convertapi-php/blob/master/lib/ConvertApi/Client.php#L24), using the [FileUpload class](https://github.com/ConvertAPI/convertapi-php/blob/master/lib/ConvertApi/FileUpload.php).
> 
> On macOS we have written a simple test script:
> 
> ```
> #!/opt/homebrew/opt/php@8.3/bin/php
> <?php
> require_once('./vendor/autoload.php');
> 
> use ConvertApi\ConvertApi;
> use ConvertApi\Client;
> 
> ConvertApi::setApiSecret('your-api-secret');
> 
> $file = 'test.docx';
> 
> $fileContents = file_get_contents($file);
> $crc = hash('crc32b', $fileContents);
> 
> echo "Original CRC: " . $crc . PHP_EOL;
> 
> $result = (new Client)->upload($file, $file);
> 
> echo "Upload url: " . $result['Url'] . PHP_EOL;
> 
> $crc = hash('crc32b', file_get_contents($result['Url']));
> echo "Uploaded CRC: " . $crc . PHP_EOL;
> ```
> 
> To replicate, in an empty folder use `composer require convertapi/convertapi-php`, copy in the above test.php script (make it executable `chmod +x test.php`, also place the [test.docx](https://github.com/ConvertAPI/convertapi-php/blob/master/examples/files/test.docx) document (provided by the repository) in the same folder.
> 
> We get the following result running the above script:
> 
> ```shell
> ➜  temp ./test.php
> Original CRC: 7937aa29
> Upload url: https://v2.convertapi.com/d/2a0r9vequ6efsm5vsa1xdzf33nahgbfc
> Uploaded CRC: 6319cecd
> ```
> 
> The [upload URL](https://v2.convertapi.com/d/2a0r9vequ6efsm5vsa1xdzf33nahgbfc) provides a corrupt file. Somewhere multiple bytes at the start of the file have been added, making it corrupt and therefore unable to convert.
> 
> Commenting out [line #30 of /ConvertApi/Client](https://github.com/ConvertAPI/convertapi-php/blob/master/lib/ConvertApi/Client.php#L30) solves the issue:
> 
> ```shell
> ➜  temp ./test.php
> Original CRC: 7937aa29
> Upload url: https://v2.convertapi.com/d/sor8qc8zufhth05hjfpa459pexgpo51w
> Uploaded CRC: 7937aa29
> ```
> 
> With [line #30 of /ConvertApi/Client](https://github.com/ConvertAPI/convertapi-php/blob/master/lib/ConvertApi/Client.php#L30) disabled, our original function worked properly again and the 5001 error disappeared.
> 
> Tested on:
> 
> PHP: 8.3.7 (attached a [phpinfo()](https://github.com/ConvertAPI/convertapi-php/files/15390350/phpinfo.log) export) macOS: 14.4.1 (23E224)
> 
> We do not believe this is an issue of the PHP Client. However, somewhere the uploaded file is getting altered/corrupt using chunked Transfer-Encoding.
> 
> Thanks in advance for looking in to this matter.

**Comment by orlu-nd** [comment]
I have done some tests.


## Test 1: Your suggested code. **_PASSES_**

```php
public function upload($file_or_resource, $fileName)
{	      
  $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, []);
  
  if (!file_exists($file_or_resource))
  {
	    throw new Error\File("Unable to open file {$file_or_resource}");
  }
  
  curl_setopt($ch, CURLOPT_POST, true);
  
  // Define the file to be uploaded
  $cfile = new \CURLFile($file_or_resource, mime_content_type($file_or_resource), basename($file_or_resource));
  
  // Attach the file to the POST fields
  $post_fields = array('file' => $cfile);
  curl_setopt($ch, CURLOPT_POSTFIELDS, $post_fields);
  
  // Set additional options
  curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
  curl_setopt($ch, CURLOPT_MAXREDIRS, 10);
  
  try
  {
	    $result = $this->execute($ch);
  }
  catch (\Exception $e)
  {
	    fclose($fp);
	    throw $e;
  }
  
  return $result;
}
```

## Test 2: Your suggested code (With Headers from original code). **_FAILS_**

```php
public function upload($file_or_resource, $fileName)
{	

  $headers = array_merge($this->defaultHeaders(), [
    'Content-Type: application/octet-stream',
    'Transfer-Encoding: chunked',
    "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
  ]);
      
  $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);
  
  if (!file_exists($file_or_resource))
  {
	    throw new Error\File("Unable to open file {$file_or_resource}");
  }
  
  curl_setopt($ch, CURLOPT_POST, true);
  
  // Define the file to be uploaded
  $cfile = new \CURLFile($file_or_resource, mime_content_type($file_or_resource), basename($file_or_resource));
  
  // Attach the file to the POST fields
  $post_fields = array('file' => $cfile);
  curl_setopt($ch, CURLOPT_POSTFIELDS, $post_fields);
  
  // Set additional options
  curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
  curl_setopt($ch, CURLOPT_MAXREDIRS, 10);
  
  try
  {
	    $result = $this->execute($ch);
  }
  catch (\Exception $e)
  {
	    fclose($fp);
	    throw $e;
  }
  
  return $result;
}
```

## Test 3: Your suggested code (With Headers, commented out 'Transfer-Encoding'). **_FAILS_**

```php
public function upload($file_or_resource, $fileName)
{	

  $headers = array_merge($this->defaultHeaders(), [
    'Content-Type: application/octet-stream',
    //'Transfer-Encoding: chunked',
    "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
  ]);
      
  $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);
  
  if (!file_exists($file_or_resource))
  {
	    throw new Error\File("Unable to open file {$file_or_resource}");
  }
  
  curl_setopt($ch, CURLOPT_POST, true);
  
  // Define the file to be uploaded
  $cfile = new \CURLFile($file_or_resource, mime_content_type($file_or_resource), basename($file_or_resource));
  
  // Attach the file to the POST fields
  $post_fields = array('file' => $cfile);
  curl_setopt($ch, CURLOPT_POSTFIELDS, $post_fields);
  
  // Set additional options
  curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
  curl_setopt($ch, CURLOPT_MAXREDIRS, 10);
  
  try
  {
	    $result = $this->execute($ch);
  }
  catch (\Exception $e)
  {
	    fclose($fp);
	    throw $e;
  }
  
  return $result;
}
```

## Test 4: Your suggested code (With Headers, commented out 'Content-Type'). **_PASSES_**

```php
public function upload($file_or_resource, $fileName)
{	

  $headers = array_merge($this->defaultHeaders(), [
    //'Content-Type: application/octet-stream',
    'Transfer-Encoding: chunked',
    "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
  ]);
      
  $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);
  
  if (!file_exists($file_or_resource))
  {
	    throw new Error\File("Unable to open file {$file_or_resource}");
  }
  
  curl_setopt($ch, CURLOPT_POST, true);
  
  // Define the file to be uploaded
  $cfile = new \CURLFile($file_or_resource, mime_content_type($file_or_resource), basename($file_or_resource));
  
  // Attach the file to the POST fields
  $post_fields = array('file' => $cfile);
  curl_setopt($ch, CURLOPT_POSTFIELDS, $post_fields);
  
  // Set additional options
  curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
  curl_setopt($ch, CURLOPT_MAXREDIRS, 10);
  
  try
  {
	    $result = $this->execute($ch);
  }
  catch (\Exception $e)
  {
	    fclose($fp);
	    throw $e;
  }
  
  return $result;
}
```

## Test 5: Your suggested code (With Headers, commented out 'Content-Type' and 'Transfer-Encoding'). **_PASSES_**

```php
public function upload($file_or_resource, $fileName)
{	

  $headers = array_merge($this->defaultHeaders(), [
    //'Content-Type: application/octet-stream',
    //'Transfer-Encoding: chunked',
    "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
  ]);
      
  $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);
  
  if (!file_exists($file_or_resource))
  {
	    throw new Error\File("Unable to open file {$file_or_resource}");
  }
  
  curl_setopt($ch, CURLOPT_POST, true);
  
  // Define the file to be uploaded
  $cfile = new \CURLFile($file_or_resource, mime_content_type($file_or_resource), basename($file_or_resource));
  
  // Attach the file to the POST fields
  $post_fields = array('file' => $cfile);
  curl_setopt($ch, CURLOPT_POSTFIELDS, $post_fields);
  
  // Set additional options
  curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
  curl_setopt($ch, CURLOPT_MAXREDIRS, 10);
  
  try
  {
	    $result = $this->execute($ch);
  }
  catch (\Exception $e)
  {
	    fclose($fp);
	    throw $e;
  }
  
  return $result;
}
```

## Test 6: Orignal code. **_FAILS_**

```php
public function upload($file_or_resource, $fileName)
{
    $headers = array_merge(
        $this->defaultHeaders(),
        [
            'Content-Type: application/octet-stream',
            'Transfer-Encoding: chunked',
            "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
        ]
    );

    $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);

    if (is_resource($file_or_resource)) {
        $fp = $file_or_resource;
    } else {
        $fp = fopen($file_or_resource, 'rb');
        
        curl_setopt($ch, CURLOPT_INFILESIZE, filesize($file_or_resource));
    }

    if (!$fp)
    {
        throw new Error\File("Unable to open file {$file_or_resource}");
    }

    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_INFILE, $fp);            

    try
    {
        $result = $this->execute($ch);
    }
    catch (\Exception $e)
    {
        fclose($fp);
        throw $e;
    }

    fclose($fp);

    return $result;
}
```

## Test 7: Orignal code (With Headers, commented out 'Transfer-Encoding'). **_FAILS_**

```php
public function upload($file_or_resource, $fileName)
{
    $headers = array_merge(
        $this->defaultHeaders(),
        [
            'Content-Type: application/octet-stream',
            //'Transfer-Encoding: chunked',
            "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
        ]
    );

    $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);

    if (is_resource($file_or_resource)) {
        $fp = $file_or_resource;
    } else {
        $fp = fopen($file_or_resource, 'rb');
        
        curl_setopt($ch, CURLOPT_INFILESIZE, filesize($file_or_resource));
    }

    if (!$fp)
    {
        throw new Error\File("Unable to open file {$file_or_resource}");
    }

    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_INFILE, $fp);            

    try
    {
        $result = $this->execute($ch);
    }
    catch (\Exception $e)
    {
        fclose($fp);
        throw $e;
    }

    fclose($fp);

    return $result;
}
```

## Test 8: Orignal code (With Headers, commented out 'Content-Type'). **_FAILS_**

```php
public function upload($file_or_resource, $fileName)
{
    $headers = array_merge(
        $this->defaultHeaders(),
        [
            //'Content-Type: application/octet-stream',
            'Transfer-Encoding: chunked',
            "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
        ]
    );

    $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);

    if (is_resource($file_or_resource)) {
        $fp = $file_or_resource;
    } else {
        $fp = fopen($file_or_resource, 'rb');
        
        curl_setopt($ch, CURLOPT_INFILESIZE, filesize($file_or_resource));
    }

    if (!$fp)
    {
        throw new Error\File("Unable to open file {$file_or_resource}");
    }

    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_INFILE, $fp);            

    try
    {
        $result = $this->execute($ch);
    }
    catch (\Exception $e)
    {
        fclose($fp);
        throw $e;
    }

    fclose($fp);

    return $result;
}
```

## Test 9: Orignal code (With Headers, commented out 'Content-Type' AND 'Transfer-Encoding'). **_PASSES_**

```php
public function upload($file_or_resource, $fileName)
{
    $headers = array_merge(
        $this->defaultHeaders(),
        [
            //'Content-Type: application/octet-stream',
            //'Transfer-Encoding: chunked',
            "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
        ]
    );

    $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);

    if (is_resource($file_or_resource)) {
        $fp = $file_or_resource;
    } else {
        $fp = fopen($file_or_resource, 'rb');
        
        curl_setopt($ch, CURLOPT_INFILESIZE, filesize($file_or_resource));
    }

    if (!$fp)
    {
        throw new Error\File("Unable to open file {$file_or_resource}");
    }

    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_INFILE, $fp);            

    try
    {
        $result = $this->execute($ch);
    }
    catch (\Exception $e)
    {
        fclose($fp);
        throw $e;
    }

    fclose($fp);

    return $result;
}
```

## Test 10: Orignal code (With Headers, commented out 'Content-Type' AND 'Transfer-Encoding' as per @firepolo suggestion). **_PASSES_**

```php
public function upload($file_or_resource, $fileName)
{
    $headers = array_merge(
        $this->defaultHeaders(),
        [
            //'Content-Type: application/octet-stream',
            'Transfer-Encoding: gzip,deflate',
            "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
        ]
    );

    $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);

    if (is_resource($file_or_resource)) {
        $fp = $file_or_resource;
    } else {
        $fp = fopen($file_or_resource, 'rb');
        
        curl_setopt($ch, CURLOPT_INFILESIZE, filesize($file_or_resource));
    }

    if (!$fp)
    {
        throw new Error\File("Unable to open file {$file_or_resource}");
    }

    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_INFILE, $fp);            

    try
    {
        $result = $this->execute($ch);
    }
    catch (\Exception $e)
    {
        fclose($fp);
        throw $e;
    }

    fclose($fp);

    return $result;
}
```

## Test 11: Orignal code (With Headers, 'Transfer-Encoding' as per @firepolo suggestion). **_PASSES_**

```php
public function upload($file_or_resource, $fileName)
{
    $headers = array_merge(
        $this->defaultHeaders(),
        [
            'Content-Type: application/octet-stream',
            'Transfer-Encoding: gzip,deflate',
            "Content-Disposition: attachment; filename*=UTF-8''" . rawurlencode($fileName),
        ]
    );

    $ch = $this->initCurl('upload', ConvertApi::$uploadTimeout, $headers);

    if (is_resource($file_or_resource)) {
        $fp = $file_or_resource;
    } else {
        $fp = fopen($file_or_resource, 'rb');
        
        curl_setopt($ch, CURLOPT_INFILESIZE, filesize($file_or_resource));
    }

    if (!$fp)
    {
        throw new Error\File("Unable to open file {$file_or_resource}");
    }

    curl_setopt($ch, CURLOPT_POST, true);
    curl_setopt($ch, CURLOPT_INFILE, $fp);            

    try
    {
        $result = $this->execute($ch);
    }
    catch (\Exception $e)
    {
        fclose($fp);
        throw $e;
    }

    fclose($fp);

    return $result;
}
```

**Comment by laurynas-convertapi** [comment]
Couldn't reproduce the issue on my machine (Ubuntu Linux + Docker), probably depends on environment. Removed content type and transfer encoding headers from file upload. Released client version 2.0.1.

**Comment by paulius-petkus** [comment]
@laurynas-convertapi Client approved that your fix is working fine. Good job.
